tensorrt_llm/_torch/auto_deploy/custom_ops/torch_rope.py

```diff
@@ -0,0 +1,187 @@
+    # Shape: [seq_len, head_dim/2]
+    emb = torch.cat((freqs, freqs), dim=-1)
+    # Ensure stable computation of sin/cos in float32
+    cos = torch.cos(emb).to(dtype=torch.float32)
+    sin = torch.sin(emb).to(dtype=torch.float32)
+
+    # Reshape for broadcasting
+    # Shape: [1, 1, seq_len, head_dim]
+    cos = cos.view(1, 1, seq_len, head_dim)
+    sin = sin.view(1, 1, seq_len, head_dim)
+
+    # Always compute in float32 for numerical stability
+    q_float = q.to(dtype=torch.float32)
+    k_float = k.to(dtype=torch.float32)
+
+    # For the even indices of the dimension
+    q_embed_even = q_float[..., 0::2]
+    q_embed_odd = q_float[..., 1::2]
+    k_embed_even = k_float[..., 0::2]
+    k_embed_odd = k_float[..., 1::2]
+
+    # Apply the rotation using the identities:
+    # q' = q * cos + rotate(q) * sin
+    # k' = k * cos + rotate(k) * sin
+    # where rotate(x) swaps the even and odd dimensions and negates the odd dimensions
+    q_rotated = torch.cat(
+        [
+            q_embed_even * cos[..., 0::2] - q_embed_odd * sin[..., 0::2],
+            q_embed_odd * cos[..., 1::2] + q_embed_even * sin[..., 1::2],
+        ],
+        dim=-1,
+    )
+
+    k_rotated = torch.cat(
+        [
+            k_embed_even * cos[..., 0::2] - k_embed_odd * sin[..., 0::2],
+            k_embed_odd * cos[..., 1::2] + k_embed_even * sin[..., 1::2],
+        ],
+        dim=-1,
+    )
+
+    # Convert back to the original dtype
+    return q_rotated.to(dtype=original_dtype), k_rotated.to(dtype=original_dtype)
+
+
+# Copied from transformers.models.llama.modeling_llama.rotate_half
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., : x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2 :]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+@torch.library.custom_op("rope::torch_apply_rope_with_explicit_cos_sin", mutates_args=())
+def torch_apply_rope_with_explicit_cos_sin(
+    q: torch.Tensor, k: torch.Tensor, cos: torch.Tensor, sin: torch.Tensor, unsqueeze_dim: int = 1
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    """
+    Reference PyTorch implementation of HF-style RoPE:
+    - Input layout: non-interleaved, [B, N, S, D] with unsqueeze_dim=1 and
+        [B, S, N, D] with unsqueeze_dim=2, default is [B, N, S, D]
+    - Frequencies are provided as separate `cos` and `sin` tensors of shape [B, S, head_dim].
+    """
+    # in HF, cos/sin tensor are passed in as x.dtype, this is to double ensure
+    cos = cos.type_as(q)
+    sin = sin.type_as(q)
+    cos = cos.unsqueeze(unsqueeze_dim)
+    sin = sin.unsqueeze(unsqueeze_dim)
+    q_embed = (q * cos) + (rotate_half(q) * sin)
+    k_embed = (k * cos) + (rotate_half(k) * sin)
+    return q_embed, k_embed
+
+
+@torch_apply_rope_with_explicit_cos_sin.register_fake
+def torch_apply_rope_with_explicit_cos_sin_fake(
+    q: torch.Tensor, k: torch.Tensor, cos: torch.Tensor, sin: torch.Tensor, unsqueeze_dim: int = 1
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    return q, k
```

Comment:
```suggestion
    return torch.empty_like(q), torch.empty_like(k)
```

It should not reference the same tensor to be consistent with the default implementation. Having the same tensor reference has implications on graph tracing and hence we wanna be consistent

Reply:
For example, in graph tracing something like `q.contiguous()` would not appear in the graph if `q` is already contiguous and hence `q.contiguous()` would just return a reference to the same tensor

Reply:
I see, thanks for explaining.